File: src/sdk/go/worker_http.go
package appmesh

import (
	"context"
	"errors"
	"fmt"
	"net/http"
	"os"
	"time"
)

// ErrProcessSuperseded indicates the server rejected the process key (HTTP 412):
// this process instance has been superseded by another and should stop fetching tasks.
var ErrProcessSuperseded = errors.New("process superseded: process key mismatch")

// ErrWorkerRejected indicates a permanent worker request error (HTTP 400).
// Retrying the same request cannot recover; the worker entry point should exit.
var ErrWorkerRejected = errors.New("worker task request permanently rejected")

// WorkerHTTPContext runs a worker-side task loop (fetch task, send result) against
// the App Mesh REST service over HTTPS.
// Requires environment variables: APP_MESH_PROCESS_KEY and APP_MESH_APPLICATION_NAME.
type WorkerHTTPContext struct {
	client *AppMeshClient
}

// NewHTTPContext creates a server-side task context over HTTP. Server endpoints
// authenticate via APP_MESH_PROCESS_KEY; the worker is not an OAuth client.
func NewHTTPContext(options Option) (*WorkerHTTPContext, error) {
	httpClient, err := NewHTTPClient(options)
	if err != nil {
		return nil, err
	}
	return &WorkerHTTPContext{client: httpClient}, nil
}
func newHTTPContextWithRequester(options Option, r Requester) (*WorkerHTTPContext, error) {
	httpClient, err := newHTTPClientWithRequester(options, r)
	if err != nil {
		return nil, err
	}
	return &WorkerHTTPContext{client: httpClient}, nil
}

// getRuntimeEnv reads and validates required runtime environment variables.
func (r *WorkerHTTPContext) getRuntimeEnv() (key, appName string, err error) {
	key = os.Getenv("APP_MESH_PROCESS_KEY")
	appName = os.Getenv("APP_MESH_APPLICATION_NAME")

	if key == "" {
		return "", "", errors.New("missing environment variable: APP_MESH_PROCESS_KEY. This must be set by App Mesh service")
	}
	if appName == "" {
		return "", "", errors.New("missing environment variable: APP_MESH_APPLICATION_NAME. This must be set by App Mesh service")
	}
	return key, appName, nil
}

// FetchTask fetches task data (payload) for the current application process.
// It retries transient failures until success with a fixed 100ms floor per attempt
// (SDKContract.md).
// It returns ErrProcessSuperseded when the server responds 412 (this process
// instance has been replaced), and ErrWorkerRejected for a permanent HTTP 400.
func (r *WorkerHTTPContext) FetchTask() (string, error) {
	return r.FetchTaskContext(context.Background())
}

// FetchTaskContext is FetchTask with a caller-supplied context controlling cancellation:
// the retry loop stops with ctx.Err() once the context is done.
func (r *WorkerHTTPContext) FetchTaskContext(ctx context.Context) (string, error) {
	key, appName, err := r.getRuntimeEnv()
	if err != nil {
		return "", err
	}

	path := "/appmesh/app/" + appName + "/task"
	headers := map[string]string{"X-AppMesh-Process-Key": key}

	// Fixed 100ms floor per attempt: sleep only the remainder if the attempt
	// finished early; otherwise retry immediately. No backoff (SDKContract.md).
	const retryFloor = 100 * time.Millisecond

	for {
		attemptStart := time.Now()
		status, body, _, err := r.client.req.SendContext(ctx, http.MethodGet, path, nil, headers, nil)
		if err != nil {
			if ctx.Err() != nil {
				return "", fmt.Errorf("fetch_task canceled: %w", ctx.Err())
			}
			logf("fetch_task request failed: %v, retrying...", err)
		} else if status == http.StatusOK {
			return string(body), nil
		} else if status == http.StatusPreconditionFailed {
			return "", ErrProcessSuperseded
		} else if status == http.StatusBadRequest {
			return "", fmt.Errorf("%w: status %d: %s", ErrWorkerRejected, status, string(body))
		} else {
			logf("fetch_task failed with status %d: %s, retrying...", status, string(body))
		}

		if remaining := retryFloor - time.Since(attemptStart); remaining > 0 {
			select {
			case <-ctx.Done():
				return "", fmt.Errorf("fetch_task canceled: %w", ctx.Err())
			case <-time.After(remaining):
			}
		}
	}
}

// SendTaskResult sends the processed result bytes back to the original invoking client via App Mesh.
func (r *WorkerHTTPContext) SendTaskResult(result string) error {
	processKey, appName, err := r.getRuntimeEnv()
	if err != nil {
		return err
	}

	path := "/appmesh/app/" + appName + "/task"
	headers := map[string]string{
		"Content-Type":          "text/plain",
		"X-AppMesh-Process-Key": processKey,
	}

	status, body, err := r.client.put(path, nil, headers, []byte(result))
	if err != nil {
		return err
	}

	if status != http.StatusOK {
		return errors.New("send_task_result failed with status " + http.StatusText(status) + ": " + string(body))
	}

	return nil
}

File: src/sdk/rust/src/worker_http.rs
// worker_http.rs
// Server SDK implementation for App Mesh (HTTP)

use bytes::Bytes;
use http::StatusCode;
use log::{error, warn};
use std::collections::HashMap;
use std::env;
use std::sync::Arc;
use std::time::Duration;
use std::time::Instant;
use tokio::sync::watch;

use crate::client_http::AppMeshClient;
use crate::error::AppMeshError;
use crate::response_ext::ResponseExt;

/// Server-side helper for applications to fetch tasks and return results.
pub struct AppMeshWorker {
    client: Arc<AppMeshClient>,
    retry_delay: Duration,
    stop_tx: watch::Sender<bool>,
}

impl AppMeshWorker {
    /// Create a new server-side task context backed by the HTTP client.
    pub fn new(
        base_url: Option<String>,
        ssl_verify: Option<String>,
        ssl_client_cert: Option<(String, String)>,
    ) -> Result<Arc<Self>, AppMeshError> {
        let client = AppMeshClient::new(base_url, ssl_verify, ssl_client_cert, None, None, false)?;
        // Server endpoints use APP_MESH_PROCESS_KEY; no JWT refresh needed.
        client.set_auto_refresh_token(false);
        Ok(AppMeshWorker::with_client(client))
    }

    fn get_runtime_env() -> Result<(String, String), AppMeshError> {
        let process_key = env::var("APP_MESH_PROCESS_KEY").map_err(|_| {
            AppMeshError::ConfigurationError(
                "Missing environment variable: APP_MESH_PROCESS_KEY. This must be set by App Mesh service.".into(),
            )
        })?;
        let app_name = env::var("APP_MESH_APPLICATION_NAME").map_err(|_| {
            AppMeshError::ConfigurationError(
                "Missing environment variable: APP_MESH_APPLICATION_NAME. This must be set by App Mesh service."
                    .into(),
            )
        })?;
        Ok((process_key, app_name))
    }

    /// Create `AppMeshWorker` with a pre-built client.
    pub fn with_client(client: Arc<AppMeshClient>) -> Arc<Self> {
        let (stop_tx, _) = watch::channel(false);
        Arc::new(Self { client, retry_delay: Duration::from_millis(100), stop_tx })
    }

    /// Cancel any in-flight or future `fetch_task` retry loop; a cancelled
    /// `fetch_task` returns `Err(AppMeshError::Cancelled)`.
    pub fn stop(&self) {
        // send_replace updates the value even with no live receivers, so a stop()
        // before/between fetches is still seen by the next fetch_task's borrow check.
        self.stop_tx.send_replace(true);
    }

    /// Fetch a task payload for the current application process.
    ///
    /// Retries until successful. If a request fails within 100ms, sleeps
    /// briefly before retrying; otherwise retries immediately.
    ///
    /// Returns `Err(AppMeshError::ProcessSuperseded)` when the daemon reports
    /// this process key is stale (HTTP 412), and `Err(AppMeshError::Cancelled)`
    /// when [`AppMeshWorker::stop`] is called.
    pub async fn fetch_task(&self) -> Result<Bytes, AppMeshError> {
        use reqwest::Method;
        use tokio::time::sleep;

        let (pkey, app_name) = Self::get_runtime_env()?;
        let path = format!("/appmesh/app/{}/task", app_name);

        let mut query = HashMap::new();
        query.insert("process_key".to_string(), pkey);

        let mut stop_rx = self.stop_tx.subscribe();
        let mut attempts: u32 = 0;
        loop {
            if *stop_rx.borrow() {
                return Err(AppMeshError::Cancelled);
            }
            let attempt_start = Instant::now();
            let result = tokio::select! {
                r = self.client.raw_request(Method::GET, &path, None, None, Some(query.clone()), false) => r,
                _ = stop_rx.changed() => return Err(AppMeshError::Cancelled),
            };
            match result {
                Ok(resp) => {
                    if resp.status() == StatusCode::OK {
                        return Ok(resp.into_bytes());
                    }
                    let status = resp.status();
                    if status == StatusCode::PRECONDITION_FAILED {
                        // Library code must never exit the host process; surface a typed error.
                        error!("Process key mismatch (412): this process has been superseded");
                        return Err(AppMeshError::ProcessSuperseded(
                            "process key mismatch (412): this process has been superseded".into(),
                        ));
                    }
                    warn!("fetch_task attempt {} failed with status {}: retrying...", attempts + 1, status);
                }
                Err(e) => {
                    warn!("fetch_task attempt {} request failed: {}: retrying...", attempts + 1, e);
                }
            }

            attempts += 1;
            if let Some(remaining) = self.retry_delay.checked_sub(attempt_start.elapsed()) {
                tokio::select! {
                    _ = sleep(remaining) => {}
                    _ = stop_rx.changed() => return Err(AppMeshError::Cancelled),
                }
            }
        }
    }

    /// Return the processed result bytes back to App Mesh so the invoker can receive them.
    pub async fn send_task_result(&self, result: &[u8]) -> Result<(), AppMeshError> {
        use reqwest::Method;

        let (pkey, app_name) = Self::get_runtime_env()?;
        let path = format!("/appmesh/app/{}/task", app_name);

        let mut query = HashMap::new();
        query.insert("process_key".to_string(), pkey);

        let resp =
            self.client.raw_request(Method::PUT, &path, Some(result), None, Some(query), false).await?;

        let status = resp.status();
        if status != StatusCode::OK {
            let text = resp.text()?;
            error!("send_task_result failed with status {}: {}", status, text);
            return Err(AppMeshError::RequestFailed { status, message: text });
        }
        Ok(())
    }
}
